File: apps/client/ios/App/App/FinpillClerkPlugin.swift
import Capacitor
import ClerkKit
import ClerkKitUI
import SwiftUI
import UIKit

@objc(FinpillClerkPlugin)
public final class FinpillClerkPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "FinpillClerkPlugin"
    public let jsName = "FinpillClerk"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "state", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "signIn", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "signOut", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getToken", returnType: CAPPluginReturnPromise),
    ]

    private var configuredKey: String?

    @objc func initialize(_ call: CAPPluginCall) {
        Task { @MainActor in
            guard let key = call.getString("publishableKey"),
                  key.hasPrefix("pk_test_") || key.hasPrefix("pk_live_") else {
                call.reject("A Clerk publishable key is required")
                return
            }
            if let configuredKey, configuredKey != key {
                call.reject("Clerk is already configured for another instance")
                return
            }
            if configuredKey == nil {
                Clerk.configure(publishableKey: key)
                configuredKey = key
            }
            call.resolve(currentState())
        }
    }

    @objc func state(_ call: CAPPluginCall) {
        Task { @MainActor in
            guard configuredKey != nil else {
                call.reject("Clerk is not configured")
                return
            }
            call.resolve(currentState())
        }
    }

    @objc func signIn(_ call: CAPPluginCall) {
        Task { @MainActor in
            guard configuredKey != nil else {
                call.reject("Clerk is not configured")
                return
            }
            guard Clerk.shared.user == nil else {
                call.resolve(currentState())
                return
            }
            guard let presenter = bridge?.viewController,
                  presenter.presentedViewController == nil else {
                call.reject("Authentication screen is unavailable")
                return
            }
            let controller = UIHostingController(
                rootView: AuthView(mode: .signIn).environment(Clerk.shared)
            )
            presenter.present(controller, animated: true) {
                call.resolve()
            }
        }
    }

    @objc func signOut(_ call: CAPPluginCall) {
        Task { @MainActor in
            guard configuredKey != nil else {
                call.reject("Clerk is not configured")
                return
            }
            do {
                try await Clerk.shared.auth.signOut()
                call.resolve(currentState())
            } catch {
                call.reject("Clerk sign-out failed")
            }
        }
    }

    @objc func getToken(_ call: CAPPluginCall) {
        Task { @MainActor in
            guard configuredKey != nil else {
                call.reject("Clerk is not configured")
                return
            }
            do {
                let token = try await Clerk.shared.session?.getToken()
                call.resolve(["token": token as Any? ?? NSNull()])
            } catch {
                call.reject("Clerk token refresh failed")
            }
        }
    }

    @MainActor
    private func currentState() -> [String: Any] {
        [
            "isLoaded": Clerk.shared.isLoaded,
            "userId": Clerk.shared.user?.id as Any? ?? NSNull(),
        ]
    }
}

File: apps/client/android/app/src/main/java/com/cuneytbozok/finpill/FinpillClerkPlugin.kt
package com.cuneytbozok.finpill

import android.content.Intent
import com.clerk.api.Clerk
import com.clerk.api.network.serialization.ClerkResult
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

@CapacitorPlugin(name = "FinpillClerk")
class FinpillClerkPlugin : Plugin() {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private var configuredKey: String? = null

    @PluginMethod
    fun initialize(call: PluginCall) {
        val key = call.getString("publishableKey")
        if (key == null || !(key.startsWith("pk_test_") || key.startsWith("pk_live_"))) {
            call.reject("A Clerk publishable key is required")
            return
        }
        if (configuredKey != null && configuredKey != key) {
            call.reject("Clerk is already configured for another instance")
            return
        }
        try {
            if (configuredKey == null) {
                Clerk.initialize(context.applicationContext, key)
                configuredKey = key
            }
            call.resolve(currentState())
        } catch (_: Exception) {
            call.reject("Clerk initialization failed")
        }
    }

    @PluginMethod
    fun state(call: PluginCall) {
        if (!requireConfigured(call)) return
        call.resolve(currentState())
    }

    @PluginMethod
    fun signIn(call: PluginCall) {
        if (!requireConfigured(call)) return
        if (!Clerk.isInitialized.value) {
            call.reject("Clerk is not ready")
            return
        }
        if (Clerk.user != null) {
            call.resolve(currentState())
            return
        }
        activity.startActivity(Intent(activity, FinpillClerkActivity::class.java))
        call.resolve()
    }

    @PluginMethod
    fun signOut(call: PluginCall) {
        if (!requireConfigured(call)) return
        scope.launch {
            try {
                when (Clerk.auth.signOut()) {
                    is ClerkResult.Success -> call.resolve(currentState())
                    is ClerkResult.Failure -> call.reject("Clerk sign-out failed")
                }
            } catch (_: Exception) {
                call.reject("Clerk sign-out failed")
            }
        }
    }

    @PluginMethod
    fun getToken(call: PluginCall) {
        if (!requireConfigured(call)) return
        if (!Clerk.isInitialized.value) {
            call.reject("Clerk is not ready")
            return
        }
        if (Clerk.user == null) {
            call.resolve(JSObject().put("token", null))
            return
        }
        scope.launch {
            try {
                when (val result = Clerk.auth.getToken()) {
                    is ClerkResult.Success -> call.resolve(JSObject().put("token", result.value))
                    is ClerkResult.Failure -> call.reject("Clerk token refresh failed")
                }
            } catch (_: Exception) {
                call.reject("Clerk token refresh failed")
            }
        }
    }

    private fun requireConfigured(call: PluginCall): Boolean {
        if (configuredKey != null) return true
        call.reject("Clerk is not configured")
        return false
    }

    private fun currentState() = JSObject()
        .put("isLoaded", Clerk.isInitialized.value)
        .put("userId", Clerk.user?.id)
}
